providers/cloudflare.go

```diff
@@ -0,0 +1,204 @@
+package providers
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/crackcomm/cloudflare"
+	"github.com/rancher/external-dns/dns"
+	"golang.org/x/net/context"
+)
+
+type CloudflareHandler struct {
+	client *cloudflare.Client
+	zone   *cloudflare.Zone
+	ctx    context.Context
+	root   string
+}
+
+func init() {
+	email := os.Getenv("CLOUDFLARE_EMAIL")
+	if len(email) == 0 {
+		logrus.Info("CLOUDFLARE_EMAIL is not set, skipping init of CloudFlare provider")
+		return
+	}
+
+	apiKey := os.Getenv("CLOUDFLARE_KEY")
+	if len(apiKey) == 0 {
+		logrus.Info("CLOUDFLARE_KEY is not set, skipping init of CloudFlare provider")
+		return
+	}
+
+	cloudflareHandler := &CloudflareHandler{}
+	if err := RegisterProvider("cloudflare", cloudflareHandler); err != nil {
+		logrus.Fatal("Could not register cloudflare provider")
+	}
+
+	cloudflareHandler.client = cloudflare.New(&cloudflare.Options{
+		Email: email,
+		Key:   apiKey,
+	})
+
+	cloudflareHandler.ctx = context.Background()
+	cloudflareHandler.root = unFqdn(dns.RootDomainName)
+
+	if err := cloudflareHandler.setZone(); err != nil {
+		logrus.Fatalf("Failed to set zone for root domain %s: %v", cloudflareHandler.root, err)
+	}
+
+	logrus.Infof("Configured %s with zone \"%s\" ", cloudflareHandler.GetName(), cloudflareHandler.root)
+}
+
+func (*CloudflareHandler) GetName() string {
+	return "Cloudflare"
+}
+
+func (c *CloudflareHandler) AddRecord(record dns.DnsRecord) error {
+	for _, rec := range record.Records {
+		r := c.prepareRecord(record)
+		r.Content = rec
+		err := c.client.Records.Create(c.ctx, r)
+		if err != nil {
+			return fmt.Errorf("CloudFlare API call has failed: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CloudflareHandler) UpdateRecord(record dns.DnsRecord) error {
+	err := c.RemoveRecord(record)
```

Comment:
Nit: when method you call returns one argument - err - that you need to evaluate, you can use construct like this:

```
if err := c.RemoveRecord(record); err != nil {
		return err
	}
```